File: solutions/php/dnd-character/1/DndCharacter.php
<?php
declare(strict_types=1);

class DndCharacter
{
    public int $strength;
    public int $dexterity;
    public int $constitution;
    public int $intelligence;
    public int $wisdom;
    public int $charisma;
    public int $hitpoints;

    /**
     * Creates a new random character.
     */
    public static function generate(): self
    {
        $character = new self();
        return $character;
    }

    public function __construct()
    {
        $this->strength = self::ability();
        $this->dexterity = self::ability();
        $this->constitution = self::ability();
        $this->intelligence = self::ability();
        $this->wisdom = self::ability();
        $this->charisma = self::ability();
        
        $modifier = self::modifier($this->constitution);
        $this->hitpoints = 10 + $modifier;
    }

    /**
     * Rolls four 6-sided dice and returns the sum of the largest three.
     */
    public static function ability(): int
    {
        $rolls = [];
        for ($i = 0; $i < 4; $i++) {
            $rolls[] = random_int(1, 6);
        }
        sort($rolls);
        // Sum highest 3 (discard the lowest)
        return array_sum(array_slice($rolls, 1));
    }

    /**
     * Calculates the modifier for a given score.
     */
    public static function modifier(int $score): int
    {
        return (int)floor(($score - 10) / 2);
    }
}

File: solutions/ruby/dnd-character/1/dnd_character.rb
class DndCharacter
  # Creamos lectores para todas las habilidades y hitpoints
  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  # Método de clase para calcular el modificador
  def self.modifier(score)
    # Restamos 10, dividimos por 2.0 para forzar flotante y usamos .floor para redondear hacia abajo
    ((score - 10) / 2.0).floor
  end

  def initialize
    @strength = roll_ability
    @dexterity = roll_ability
    @constitution = roll_ability
    @intelligence = roll_ability
    @wisdom = roll_ability
    @charisma = roll_ability
    
    # HP = 10 + modificador de constitución
    @hitpoints = 10 + self.class.modifier(@constitution)
  end

  private

  def roll_ability
    # 1. Generamos 4 números aleatorios entre 1 y 6
    # 2. Los ordenamos de menor a mayor
    # 3. Quitamos el primero (el más bajo) con .last(3) o .drop(1)
    # 4. Sumamos los restantes
    Array.new(4) { rand(1..6) }.sort.last(3).sum
  end
end